enhancements/microshift/feature-csi-snapshots.md

```diff
@@ -0,0 +1,572 @@
+> NOTE: `volumeSnapshot` and `volumeSnapshotContent` instances will have finalizers set on them to control the deletion 
+> sequence.
+
+1. The user deletes the `volumeSnapshot` object. Finalizers on this object prevent immediate deletion. 
+
+_If the reclaimPolicy is **Delete**:_
+
+2. The `CSI snapshot controller` detects the deletion event and deletes the bound `volumeSnapshotContent` 
+   object. Finalizers on this object prevent immediate deletion.
+3. The _CSI external-snapshotter_ detects the `volumeSnapshotContent` deletion event.  It calls the `DeleteSnapshot` CSI 
+   gRPC.
+4. LVMS executes the `DeleteSnapshot` process.  The LVM snapshot is destroyed and the process returns to the caller.
+5. The _CSI external-snapshotter_ removes the finalizer on the `VolumeSnapshotContent`, allowing it to be garbage collected
+6. The _CSI snapshot controller_ removes the finalizer on the `VolumeSnapshot`, allowing it to be garbage collected
+
+_If the reclaimPolicy is **Retain**:_
+
+2. The `CSI snapshot controller` detects the deletion event. It removes finalizers on `volumeSnapshot` and 
+   `volumeSnapshotContent` objects.
+3. The `volumeSnapshot` is garbage collected.
+4. The `volumeSnapshotContent` and backing LVM thin volume are retained.  A user can then delete the backing storage 
+   by deleting the `volumeSnapshotContent`.  
+
+#### Restore
+
+Restoring is the process of creating a new lvm volume from a snapshot that is populated with data from the snapshot. 
+This can be achieved two ways via the `spec.dataSource` PVC field.  
+
+_Prerequisites_
+
+* A `storageClass` is deployed automatically
+* A `volumeSnapshotClass` is deployed automatically
+* A `volumeSnapshot` is bound to a `volumeSnapshotContent` and `readyToUse` is `true
+
+_Workflow_
+
+1. The user creates a PVC with the following fields:
+   - `spec.storageClassName`: must be the same storage class that provisioned the source volume
+   - `spec.dataSource.name`: name of the source `volumeSnapshot`
+   - `spec.dataSource.kind`: VolumeSnapshot
+   - `spec.dataSource.apiGroup`: snapshot.storage.k8s.io
+2. The user creates a Pod to consume the PVC. Required because LVMS only supports `WaitForFirstConsumer` provisioning.
+3. The _CSI external provisioner_ sidecar, already integrated into the _topolvm-controller_, detects the PVC request,
+   and makes a `CreateVolume` gRPC call to LVMS. 
+4. _topolvm-controller_ creates a `logicalVolume.topolvm.io` CRD with the same values of the source `logicalVolume.
+   topolvm.io` instance. This ensures the new volume is created on the same node as the source.
+5. _topolvm-node_ creates a new thin-volume from the volume snapshot.
+6. _topolvm-controller_ returns the success status to the `CreateVolume` gRPC caller.
+7. The _CSI external provisioner_ creates a `persistentVolume` to track the new thin-volume
+8. The _CSI external provisioner_ binds the PVC and PV together
+9. The workload attaches the new volume and starts
+
+
+#### Volume Cloning
```

Comment:
Technically, you don't need Kubernetes snapshots (CRDs, external-snapshotter sidecar, snapshot-controller), all the handling below is implemented in the external-provisioner and topolvm CSI driver that you already have.

Reply:
Does that mean we don't need to bring in some of the components mentioned in this doc? What API would a user use to trigger a snapshot?

Reply:
In Kubernetes, user asks directly in a PVC to provision a new volume as a clone of another PVC in the same namespace.
https://github.com/kubernetes/enhancements/blob/26c0a72b9b216dc86a6e02bcdf96eb9f6afce0d3/keps/sig-storage/989-extend-datasource/README.md?plain=1#L71-L73

It's up to the CSI driver what it does with such request, in LVM case it creates a snapshot under the hood. But the snapshot is not exposed in Kubernetes API.

Reply:
OK, the user we're working with wants to integrate with a specific backup tool. @copejon , let's see if we can find out if that tool can use the approach @jsafrane describes here.

Reply:
I'll pull the cloning section since it's out of scope.  @dhellmann I'll update the microshift docs with it instead.